File: packages/sdk-js/src/hash.ts
/**
 * MurmurHash3, x86 32-bit variant. Austin Appleby's reference algorithm,
 * public domain: https://github.com/aappleby/smhasher
 *
 * Implemented here rather than imported because (a) the SDK has a 15 KB budget
 * and (b) every SDK in every language must produce identical output for
 * identical input. A 40-line function you can read is easier to keep identical
 * than a dependency you cannot.
 *
 * Input is hashed as UTF-8 bytes. That matters for non-ASCII flag keys and
 * user IDs: `murmurhash-js` on npm hashes UTF-16 code units and would disagree
 * with Python's `bytes` on anything outside ASCII.
 */
export function murmurhash3_x86_32(input: string, seed = 0): number {
  const data = new TextEncoder().encode(input);
  const len = data.length;
  const nblocks = len >>> 2;

  const c1 = 0xcc9e2d51;
  const c2 = 0x1b873593;

  let h1 = seed >>> 0;

  // Body: 4 bytes at a time, little-endian.
  for (let i = 0; i < nblocks; i++) {
    const p = i * 4;
    let k1 = (data[p] | (data[p + 1] << 8) | (data[p + 2] << 16) | (data[p + 3] << 24)) >>> 0;

    k1 = Math.imul(k1, c1);
    k1 = (k1 << 15) | (k1 >>> 17);
    k1 = Math.imul(k1, c2);

    h1 ^= k1;
    h1 = (h1 << 13) | (h1 >>> 19);
    h1 = (Math.imul(h1, 5) + 0xe6546b64) >>> 0;
  }

  // Tail: the 0-3 bytes that did not fill a block.
  const tail = nblocks * 4;
  let k1 = 0;
  switch (len & 3) {
    case 3:
      k1 ^= data[tail + 2] << 16;
    // falls through
    case 2:
      k1 ^= data[tail + 1] << 8;
    // falls through
    case 1:
      k1 ^= data[tail];
      k1 = Math.imul(k1, c1);
      k1 = (k1 << 15) | (k1 >>> 17);
      k1 = Math.imul(k1, c2);
      h1 ^= k1;
  }

  // Finalization: force all bits of the hash to avalanche.
  h1 ^= len;
  h1 ^= h1 >>> 16;
  h1 = Math.imul(h1, 0x85ebca6b);
  h1 ^= h1 >>> 13;
  h1 = Math.imul(h1, 0xc2b2ae35);
  h1 ^= h1 >>> 16;

  return h1 >>> 0;
}

/**
 * Which of 100 buckets this user lands in for this flag. 0-99 inclusive.
 * Seed is 0 and the separator is ":" - both are part of the cross-SDK contract
 * and can never change.
 */
export function bucket(flagKey: string, userId: string): number {
  return murmurhash3_x86_32(`${flagKey}:${userId}`, 0) % 100;
}

File: packages/sdk-python/flagrship/hash.py
"""MurmurHash3, x86 32-bit variant.

Austin Appleby's reference algorithm, public domain:
https://github.com/aappleby/smhasher

Implemented here rather than imported so that every SDK, in every language,
produces identical output for identical input. This file must stay
byte-for-byte compatible with packages/sdk-js/src/hash.ts.

Input is hashed as UTF-8 bytes. Python's `str.encode()` and JavaScript's
`TextEncoder` agree on that; a hash over code points would not.

Python integers are arbitrary-precision, so every multiply and rotate is
masked to 32 bits explicitly. Miss one mask and the hash silently diverges
from the JS SDK on inputs longer than a few bytes.
"""

_MASK = 0xFFFFFFFF
_C1 = 0xCC9E2D51
_C2 = 0x1B873593


def _rotl32(x: int, r: int) -> int:
    return ((x << r) | (x >> (32 - r))) & _MASK


def murmurhash3_x86_32(data: str | bytes, seed: int = 0) -> int:
    """Return the 32-bit MurmurHash3 of `data` as an unsigned int."""
    if isinstance(data, str):
        data = data.encode("utf-8")

    length = len(data)
    nblocks = length // 4
    h1 = seed & _MASK

    # Body: 4 bytes at a time, little-endian.
    for i in range(nblocks):
        p = i * 4
        k1 = data[p] | (data[p + 1] << 8) | (data[p + 2] << 16) | (data[p + 3] << 24)

        k1 = (k1 * _C1) & _MASK
        k1 = _rotl32(k1, 15)
        k1 = (k1 * _C2) & _MASK

        h1 ^= k1
        h1 = _rotl32(h1, 13)
        h1 = (h1 * 5 + 0xE6546B64) & _MASK

    # Tail: the 0-3 bytes that did not fill a block.
    tail = nblocks * 4
    k1 = 0
    remaining = length & 3
    if remaining == 3:
        k1 ^= data[tail + 2] << 16
    if remaining >= 2:
        k1 ^= data[tail + 1] << 8
    if remaining >= 1:
        k1 ^= data[tail]
        k1 = (k1 * _C1) & _MASK
        k1 = _rotl32(k1, 15)
        k1 = (k1 * _C2) & _MASK
        h1 ^= k1

    # Finalization: force all bits of the hash to avalanche.
    h1 ^= length
    h1 ^= h1 >> 16
    h1 = (h1 * 0x85EBCA6B) & _MASK
    h1 ^= h1 >> 13
    h1 = (h1 * 0xC2B2AE35) & _MASK
    h1 ^= h1 >> 16

    return h1


def bucket(flag_key: str, user_id: str) -> int:
    """Which of 100 buckets this user lands in for this flag. 0-99 inclusive.

    Seed 0 and the ":" separator are part of the cross-SDK contract and can
    never change.
    """
    return murmurhash3_x86_32(f"{flag_key}:{user_id}", 0) % 100
